### PR TITLE
Abort multipart upload 

### PR DESCRIPTION
**Pull request by lynnliu030**
Implement `abort_multipart_upload`. 

Also resolves the following so that it pass all juiceFS obj store [functionality check](https://juicefs.com/docs/community/performance_evaluation_guide/): 
* https://github.com/skyplane-project/skystore/issues/38
* List objects: deal with `start_after` and special key encoding

### CODE REVIEW

File: s3-proxy/src/client_impls/azure.rs [hunk 1]
```diff
@@ -458,4 +458,60 @@ impl ObjectStoreClient for AzureObjectStoreClient {
...
+                        }
+                    })
+                    .map(|block_with_size| block_with_size.block_list_type.clone())
+                    .collect();
+
+                let block_list = BlockList {
+                    blocks: blocks_to_retain,
+                };
+
+                // Not include the blocks with the matching prefix
```
Comment (simon-mo) [inline s3-proxy/src/client_impls/azure.rs:498]:
```suggestion
                // Not include the blocks with the matching prefix. Azure doesn't have a delete block API. 
```

File: s3-proxy/src/skyproxy.rs [hunk 2]
```diff
@@ -399,19 +394,28 @@ impl S3 for SkyProxy {
             Ok(resp) => {
                 let mut objects: Vec<Object> = Vec::new();
 
+                // to deal with special key
```
Comment (simon-mo) [inline s3-proxy/src/skyproxy.rs:397]: is this all we need to deal with special keys? is this the best place to implement it? I would imagine we now have to adopt such encoding/decoding in a lot of places? What does each cloud accept? Are we encoding it in the database as well?

File: s3-proxy/justfile [hunk 3]
```diff
@@ -38,8 +38,8 @@ run-registry:
 
 install-local-s3:
   cargo install --force \
-    --git https://github.com/simon-mo/s3s.git \
-    --rev 579b49e10c8bc3662786d57dbe718f20de512a2c \
+    --git https://github.com/lynnliu030/s3s.git \
+    --rev 5b84de9f8b7bc9434f0054e91b3654a6ee876317 \
```
[the remark this answers is not in the document]
Comment (lynnliu030) [reply s3-proxy/justfile:42]: I did already: https://github.com/Nugine/s3s/commit/ed3a7e6e6e6ded870bffb8ba4fa932b793f1bac6 it's here!

[the remark this answers is not in the document]
Comment (simon-mo) [reply s3-proxy/justfile:42]: In that case, let's just use the commit from Nugine/s3s?

**Review by simon-mo — APPROVED** [review]
LGTM, please address other comments before merge

File: s3-proxy/src/client_impls/azure.rs [hunk 4]
```diff
@@ -458,4 +458,60 @@ impl ObjectStoreClient for AzureObjectStoreClient {
...
+                        if let BlobBlockType::Uncommitted(block_id) =
+                            &block_with_size.block_list_type
+                        {
+                            let block_id_str =
+                                String::from_utf8(block_id.bytes().to_vec()).unwrap_or_default();
+                            !block_id_str.starts_with(&format!("{}-", upload_id))
+                        } else {
+                            true
+                        }
+                    })
```
[the remark this answers is not in the document]
Comment (lynnliu030) [reply s3-proxy/src/client_impls/azure.rs]: Change here: https://github.com/skyplane-project/skystore/pull/37/files#diff-f3ff6aa8d7157628d805c14ca04286be24732b22b922718388e0bdee9fdafa90R472-R476 to fetch only uncommitted blocks

File: s3-proxy/src/skyproxy.rs (hunk 2 above)
Comment (lynnliu030) [reply s3-proxy/src/skyproxy.rs]: This is not related to the actual encoding of file, but just a way to encode the object keys in the response according to the encoding method (e.g. URL): https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/s3/model/EncodingType.html

**Pull request merged**